### ISSUE TITLE
[Question] How to properly dispose a window and all its content

### ISSUE THREAD

**Issue by castben** [ceccopierangiolieugenio/pyTermTk#62]
Hi,

Sorry if this a dumb question, but what is the best way to close properly a window and destroy all its content?

I've been searching for an example but I've not found anything...

I've tried to use .close() method from TTKWindow() class but not sure what I'm missing as it do not close window... only thing I'm able to do is hide and show window... but actually I wanted to destroy it... unless 

Thanks

**Comment by ceccopierangiolieugenio** [comment]
So far I use the python garbage collector to destroy the object, in the "close" routine I try to dereference all the pointers:

https://github.com/ceccopierangiolieugenio/pyTermTk/blob/ae13f0749c5912e00283f794a4e234e9627fb9d6/TermTk/TTkWidgets/widget.py#L566-L572

This should release all the resources as long as the widget is not references somewhere else.

I implemented a __DEL__ routine long ago just to test which widget was destroyed and I noticed that after the close the object was released, but I never made a deep investigation on it or tried to stress and profiling it,

I need to add some extra routine to force the release of all the signals and slots as well in the close.

Anyway I would be happy to try to investigate this a bit more.

About your issue with the TTkWindow,
Are you able to share the code?
I may have miss something since I don't remember I ever tried to close a Window.

**Comment by ceccopierangiolieugenio** [comment]
The color or file picker use "close()" to close the window:
https://github.com/ceccopierangiolieugenio/pyTermTk/blob/777ce187226b9a894b9e1b4b708297dfb27d1ac8/TermTk/TTkWidgets/TTkPickers/colorpicker.py#L279-L290

**Comment by castben** [comment]
Hi Thank you for let me know.

Let me doublecheck I think it's probably some of internal widget are being referenced somewhere else.

I've included in my code an "Inventory" of widgets that basically give to me access to some widgets to get their values.

As I'm working with forms, I need to access easily this, it may be possible that this method is interfering with close() method and it do not close window properly.

The behaviour I have is, when I ask kindow to close it just blinks and clean up all its content... but everityhing still there and then start misbehaving, let me make sure as you suggest to do not reference any child widget in that window, so I will make sure to release any reference and then try to close window.

About code here is where this happens (well where I'm trying to use this close method)


```
def loadLab():
    """
    Method that will list current saved labs, to be deployed, allowing user to select previous saved lab
    TODO: Still dealing method to "destroy a window", at the moment window is being "RE-CREATED" as brand new,
    TODO: and previous window will keep around, I need to find a way to destroy previous window OR un-hide it if window
    TODO: was hiding previously
    :return:
    """
    global rootWindow
    list_plans = LabInfra.list_plans()
    root_layout = ttk.TTkVBoxLayout()
    button_layout = ttk.TTkGridLayout()
    load_lab_window = ttk.TTkWindow(parent=rootWindow, pos=(46, 2), size=(80, 20), title="Load lab plan",
                                    border=True, layout=root_layout)

    listlabs = ttk.TTkList(name="LabPlan", size=(80, 20))

    headers = {
        'Customer name': 25,
        'Ticket number': 15,
        'Brief description': 25,
        'File name': 25
    }
    head_line = ""
    for each_h in headers:
        head_line += f"{each_h:{headers[each_h]}} "
    listlabs.addItem(head_line)
    listlabs.addItem("-"*100)
    for each_line in list_plans:
        head_line = ""
        for each_h in headers:
            head_line += f"{list_plans[each_line][each_h]:{headers[each_h]}}|"
        head_line.strip('|')
        listlabs.addItem(head_line)

    load = ttk.TTkButton("Load", text="Load")
    close = ttk.TTkButton("Close", text="Close")

    def load_selected_lab():
        """
        Load selected lab into
        :return:
        """
        global current_lab
        selection = listlabs.selectedItems()[0].text.split('|')
        ttk.TTkLog.info(f"Selected lab: {selection[3]}")
        loaded_lab = f"{selection[0].strip()}:{selection[1].strip()}"
        mwin = Tui.get_widget("Menu")                                   <-------- This is coming from my internal "Inventory" of widgets
        mwin.setTitle(f"Menu-[{loaded_lab}]")
        current_lab = LabInfra()
        current_lab.load_plan(selection[3])

    load.clicked.connect(lambda: load_selected_lab())

    button_layout.addWidget(load)
    button_layout.addWidget(close)
    root_layout.addWidget(listlabs)
    root_layout.addItem(button_layout)

    close.clicked.connect(lambda: load_lab_window.close())

```

And this is part of this inventory helper class:
```
class Tui:
    """
     A helper class to build forms quickly, it provides a quick way to gather information from widgets, also
     allows to group widgets under same name.
    """
    widget_inventory = []
    widget_inventory_index = {}
    widget_inventory_groups = []
    widget_group_exclusions = []

    def __init__(self):
        """

        """
        self.form_data = []
        self.form_frame = ttk.TTkLayout()
        self.starting_row = 0

    def set_form_frame(self, frame):
        """
        Set actual form frame to attach widgets to
        :return:
        """
        self.form_frame = frame

    def add_widget(self, label=None, widget=None, group=None, widget_list=None):
        """
        Add given widget to the form. This action will also add this field to an internal reference inventory,
        this inventory will be filled with all widgets that may contain a value; also it is possible to
        group several widgets under same group name, so they can be accessed easily.

        :param label: Label that this widget will have
        :param widget: widget itself to be added
        :param group: this will group all widgets with the same group name for easy access
        :return:
        """

        if not widget_list:
            self.form_data.append({label: widget})
            if not label:
                widget_name = widget._name
            else:
                if isinstance(label, str):
                    widget_name = label.strip()
                else:
                    widget_name = label._name
            Tui.add_to_inventory(widget, widget_name, group)

        if widget_list:
            for each_item in widget_list:
                Tui.add_to_inventory(each_item, each_item._name, group)

            self.form_data.append(widget_list)

    @staticmethod
    def add_to_inventory(widget, label=None, group=None):
        """
        Used to add given widget into internal widget inventory -- given item will be saved into a list, a spcial
        index it is being build
        :param label: name for the widget, if none is given will try to get it actual widget name
        :param widget: widget that want to save
        :param group: optional group name you want this widget to belong to.
        :return:
        """
        # Get widget name
        widget_name = widget._name
        # Get actual memory address reference
        wid = hex(id(widget))
        # calculate which is last available position for this inventory
        pos = len(Tui.widget_inventory)
        if label:
            if isinstance(label, str):
                wname = label
            else:
                wname = widget_name
        else:
            wname = "Unknown"

        # Save into inventory list given object
        Tui.widget_inventory.append(widget)
        # Create "Data" reference that points to inventory, effectively working as an index
        Tui.widget_inventory_index[wid] = {
            "widget-name": label,
            "label-name": wname,
            "index-pos": pos,
            "group": group
        }

        if group and group not in Tui.widget_inventory_groups:
            Tui.widget_inventory_groups.append(group)
```


Basically window will looks like this:
![image](https://user-images.githubusercontent.com/56597585/198566013-25076d28-5e81-415a-beeb-85c642bf9811.png)

I'm still deciding if  a list widget is the best option for this (as you can see I've added header also as an item which I did not like at all, but well as I'm still learning I may redo this to take out headers and put them as labels outside... but this is the best I came with for now as I need to speed up my project :), so I may change this interface later...

Thank you I really appreciate your help.

**Comment by ceccopierangiolieugenio** [comment]
I managed to reproduce the issue [[you can try it online](https://ceccopierangiolieugenio.github.io/pyTermTk/sandbox/sandbox.html?fileUri=https://gist.githubusercontent.com/ceccopierangiolieugenio/91b4f50fca629569fc2a31fd689c2aec/raw/9f68d6e0ac9378ec6bcab89f28e2f8e193784de6/issue_62_001.py)]
There is a bug when I dereference the resources,
Having a linked list with the widgets is not an issue as long as they are not attached to any layout.
I have a quick fix but I can still see a glitch that I need to investigate.
If you want it to be released I can push it.

**Comment by ceccopierangiolieugenio** [comment]
Should be fixed in the latest release,
you can try it in the [Sandbox](https://ceccopierangiolieugenio.github.io/pyTermTk/sandbox/sandbox.html?fileUri=https://gist.githubusercontent.com/ceccopierangiolieugenio/91b4f50fca629569fc2a31fd689c2aec/raw/9f68d6e0ac9378ec6bcab89f28e2f8e193784de6/issue_62_001.py)

**Comment by castben** [comment]
Hi
Thank you for your always welcomed updates!

Cool I will update my PyTermTk Framework and will report back. Will follow your advices.

Regards!

**Comment by castben** [comment]
Ok I just updated my environment with your last push, and modified close call and now it works! 

Thank you very much for your always quick solutions! you can consider this issue resolved from my side.

Regards,

**Comment by ceccopierangiolieugenio** [comment]
thanks
